### PR TITLE
Fix bugs on leadership calendar

### PR DESCRIPTION
**Pull request by kurtrwall**
Related to [GHE]/CFGOV/platform/issues/131

There are template errors and mishandled PDF logic that is fixed with this.

## Additions

- Macros in the leadership-calendar-table.html template for modularity and clarity.

## Changes

- Failed PDFreactor calls return the print template rendered instead of just passing.
- Fixed print template macro call
- Refactored templates and view logic
- Fixed markup for "checked" calendars

## Testing

- Go to leadership calendar
- Filter on a calendar
- Filter on dates
- Click Download PDF button on the bottom controls form
- If PDFreactor is set up then it should return a pdf. Otherwise, it should direct you to the print page with all the events.

## Review

- @rosskarchner 
- @willbarton 
- @Scotchester 
- @chosak 
- @virginiacc 
- @contolini 

## Todos

- Tests

## Checklist

* [ ] Changes are limited to a single goal (no scope creep)
* [ ] Code can be automatically merged (no conflicts)
* [ ] Code follows the standards laid out in the [front end playbook](https://github.com/cfpb/front-end)
* [ ] Passes all existing automated tests
* [ ] New functions include new tests
* [ ] New functions are documented (with a description, list of inputs, and expected output)
* [ ] Placeholder code is flagged
* [ ] Visually tested in supported browsers and devices
* [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)

### CODE REVIEW

**Comment by chosak** [comment]
It would be good if there were some tests around this functionality.

What's the rationale behind rendering the regular page if something goes wrong with PDFreactor, rather than returning a (temporary) redirect to the page? This adds a `/pdf/` to the URL that then sticks around as you go through the pages, which is strange. If a user wasn't sure what happened, and tried again, you then get `/pdf/pdf` in the URL, which returns a 404. If we did a redirect instead, we could use something like the messages framework to add some kind of error message to be shown on the page letting users know that something went wrong.

**Comment by kurtw** [comment]
@chosak I don't see that happening. It doesn't use the "regular" page, it uses a different template. This is what I see if pdf reactor fails.

screen shot 2016-07-28 at 10 23 55 am

**Comment by kurtw** [comment]
@chosak ahh I see what you mean. If PDFreactor library isn't installed then it will use the regular template. I'll change it real quick. Standby.

**Comment by chosak** [comment]
As a user my expectation when clicking "Download PDF" is that I'll get a PDF download, or, if something goes wrong, that I'll see some kind of error or explanation. Showing the print template might be better than nothing but it's still confusing to the user especially if there's no error message.

File: cfgov/cal/views.py [hunk 1]
```diff
@@ -150,4 +152,4 @@ def pdf_response(request, context):
         response['Content-Disposition'] = 'attachment; filename=cfpb-leadership.pdf'
         return response
     except (urllib2.HTTPError, urllib2.URLError, BadStatusLine):
-        pass
+        return render(request, template_name, context)
```
Comment (vccabral) [inline cfgov/cal/views.py]: should we be raising an exception? raise500 etc.

File: cfgov/cal/views.py [hunk 2]
```diff
@@ -122,8 +122,10 @@ def display(request, pdf=False):
             'range_end':range_end,
         })
 
-        if pdf and PDFreactor:
-            return pdf_response(request, context)
+        if pdf:
+            template_name = 'about-us/the-bureau/leadership-calendar/print/index.html'
```
Comment (vccabral) [inline cfgov/cal/views.py]: line above is not used in one of the branches below.

File: cfgov/jinja2/v1/_includes/leadership-calendar-table.html [hunk 3]
```diff
@@ -1,33 +1,59 @@
...
-        <tr>
-            <th colspan="3">
-                {{ time.render(day, {'date':true}) }}
-            </th>
-        </tr>
-    </thead>
-    <tbody>
-    {% for event in paged_events|sort(attribute='dtstart') %}
+{% macro render_events(calendar_events, day, options) %}
+    {% if day.date is defined %}
```
Comment (vccabral) [inline cfgov/jinja2/v1/_includes/leadership-calendar-table.html]: what is this template magic?

File: cfgov/jinja2/v1/_includes/leadership-calendar-table.html [hunk 4]
```diff
@@ -1,33 +1,59 @@
...
+        {% set day = day.date() %}
+    {% endif %}
+    {% for event in calendar_events|sort(attribute='dtstart') %}
```
Comment (vccabral) [inline cfgov/jinja2/v1/_includes/leadership-calendar-table.html]: Can you sort before you get here? This is considered poor form to sort in the template.

File: cfgov/cal/views.py (hunk 1 above)
Comment (rosskarchner) [inline cfgov/cal/views.py]: @chosak @vccabral that may be a good idea-- but what @kurtw has done here matches the current behavior. If we want to change that, it should be a seperate PR.

File: cfgov/jinja2/v1/_includes/leadership-calendar-table.html (hunk 3 above)
Comment (rosskarchner) [inline cfgov/jinja2/v1/_includes/leadership-calendar-table.html]: @vccabral It's a Jinja2 feature: http://jinja.pocoo.org/docs/dev/templates/#tests

**Comment by kurtw** [comment]
@chosak I tend to agree with that sentiment. This workflow is just what we've been using. If we want to change it, I think it would be in another PR since the point of this one is to fix bugs.

File: cfgov/cal/views.py (hunk 2 above)
Comment (kurtw) [inline cfgov/cal/views.py]: I don't understand this.

**Comment by chosak** [comment]
@kurtw @rosskarchner thanks for the explanation. It'd be good if the current behavior was encoded in unit tests. Currently the `cal` app has no tests.

**Comment by willbarton** [comment]
> @kurtw @rosskarchner thanks for the explanation. It'd be good if the current behavior was encoded in unit tests. Currently the cal app has no tests.

At the very least, can we include tests that verify the bugs this PR fixes are fixed?

**Comment by kurtw** [comment]
I will include tests shortly after my meeting (1pm today).

File: cfgov/cal/views.py (hunk 2 above)
Comment (vccabral) [inline cfgov/cal/views.py]: The template_name is not used if PDFreactor is true, you can move this line below the condition if desired.

Comment (vccabral) [inline cfgov/cal/views.py]: Or you can tri-branch the return.

File: cfgov/cal/views.py (hunk 1 above)
Comment (vccabral) [inline cfgov/cal/views.py]: 👍

File: cfgov/jinja2/v1/_includes/leadership-calendar-table.html (hunk 3 above)
Comment (vccabral) [inline cfgov/jinja2/v1/_includes/leadership-calendar-table.html]: 👍

**Comment by rosskarchner** [comment]
@kurtrwall are those tests still happening?

**Comment by kurtrwall** [comment]
@rosskarchner pushed up up and ready for a `tox`

**Comment by rosskarchner** [comment]
👍 @vccabral 's feedback is worth noting, but I definitely want to get this into this weeks release to beta.

File: cfgov/cal/test_views.py [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+from django.http import HttpResponse
+from django.test import TestCase
+from django.test.client import RequestFactory
+
+from .views import display, set_cal_events_context, set_pagination_context, \
+    get_calendar_events_query, pdf_response
+
+
+class TestCalendarEvents(TestCase):
+
```
Comment (chosak) [inline cfgov/cal/test_views.py]: pep8: extra blank line

File: cfgov/cal/test_views.py [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        mock_form_class.return_value = mock_form
+        display(self.request)
+        assert mock_form.is_valid.called
+
+    @mock.patch('cal.views.render')
+    @mock.patch('cal.views.CalendarFilterForm')
+    @mock.patch('cal.views.pdf_response')
+    @mock.patch('cal.views.set_cal_events_context')
+    @mock.patch('cal.views.PDFreactor')
+    def test_display_calls_pdf_response(self, mock_PDFreactor, mock_set_cal_events_context, mock_pdf_response, mock_form_class, mock_render):
```
Comment (chosak) [inline cfgov/cal/test_views.py]: pep8: max line length

File: cfgov/cal/views.py [hunk 7]
```diff
@@ -68,65 +71,87 @@ def display(request, pdf=False):
...
+        calendars = form.cleaned_data['filter_calendar']
+        cal_q &= Q(('calendar__in', calendars))
+
+    if form.cleaned_data.get('filter_range_date_gte'):
+        gte = form.cleaned_data['filter_range_date_gte']
+        cal_q &= Q(('dtstart__gte', gte))
+
+    if form.cleaned_data.get('filter_range_date_lte'):
+        # adding timedelta(days=1) makes the end of the range inclusive
+        lte = form.cleaned_data['filter_range_date_lte']+ timedelta(days=1)
```
Comment (chosak) [inline cfgov/cal/views.py]: pep8: missing space before `+`

**Comment by chosak** [comment]
There are some Python coding standard things that it would be good to clean up. Besides that, validated functionality works, and thanks for adding tests. 👍

**Comment by rosskarchner** [comment]
OK-- there are some TODO's but let's merge this.

**Pull request merged**